### ISSUE TITLE
Webmin fails after using update-from-repo.sh

### ISSUE THREAD

**Issue by Admiral-Noisy-Bottom** [webmin/webmin#1252]
I've made a  huge mistake and ran update-from-repo.sh after it was suggested when updating the Authentic Theme.

Now my Webmin will not load at all. Usermin does load but I cannot access anything via the web interface.

I recognise that I've dug a hole for myself by running the update and accept full responsibility for ignoring the warning telling me "this might break your installation".

My question is, what are the ramifications of running /usr/libexec/setup.sh again? My system has 6 domains, postfix and all the stuff which has been running flawlessly for 12 months. If I run setup.sh again will is change all that or just reinstall Webmin.

I intend to backup /etc/webmin /etc/usermin before running the setup.

Any help would be appreciated.

Craig

**Comment by swelljoe** [comment]
> My question is, what are the ramifications of running /usr/libexec/setup.sh again?

Don't dig the hole deeper! That won't do anything good.

What error do you get when you try to start/restart Webmin?

**Comment by Admiral-Noisy-Bottom** [comment]
Thanks for such a fast reply.

My original error was about PAM not being installed, which was weird. After some research I found I also needed to run yum install pam-devel, which resolved that issue.

After running update-from-repo.sh again the error found in miniserv.error was

` Perl execution failed : Wide character in syswrite at /usr/libexec/webmin/miniserv.pl line 3140`

At one point after the first attempt webmin did start and when I tried to load the web interface I was asked to log in as usual. When I logged in the page went blank. 

There are no other errors in miniserv.log.

**Comment by Admiral-Noisy-Bottom** [comment]
In addition to my previous reply, when I run systemctl status webmin.service I get the following;

`webmin.service - LSB: web-based administration interface for Unix systems
   Loaded: loaded (/etc/rc.d/init.d/webmin; bad; vendor preset: disabled)
   Active: failed (Result: exit-code) since Wed 2020-06-24 09:22:07 AWST; 23min ago
     Docs: man:systemd-sysv-generator(8)
  Process: 8154 ExecStop=/etc/rc.d/init.d/webmin stop (code=exited, status=0/SUCCESS)
  Process: 15143 ExecStart=/etc/rc.d/init.d/webmin start (code=exited, status=25)

Jun 24 09:22:06 mail systemd[1]: Starting LSB: web-based administration interface for Unix systems...
Jun 24 09:22:07 mail systemd[1]: webmin.service: control process exited, code=exited status=25
Jun 24 09:22:07 mail systemd[1]: Failed to start LSB: web-based administration interface for Unix systems.
Jun 24 09:22:07 mail systemd[1]: Unit webmin.service entered failed state.
Jun 24 09:22:07 mail systemd[1]: webmin.service failed.`

**Comment by swelljoe** [comment]
PAM error is a red herring. It won't prevent Webmin from starting (it's wrapped in a try/catch eval) and is not needed by most people.

When the page goes blank is what we want to see errors for. There was nothing new in miniserv.error at that point? Also check `miniserv.log` (access log) and `webmin.log` (actions log). It *should* be in .error, but if it's just not finding some resource it could be a 404 or something in the access log.

Since there was also a theme update, check the developer console in your browser for errors.

**Comment by Admiral-Noisy-Bottom** [comment]
Thanks, I'll take a look and get back to you shortly.

**Comment by Admiral-Noisy-Bottom** [comment]
Ok, I checked it out. Interestingly, after rebooting the server, Webmin started and systemctl status webmin.service shows it running and ok.

Still blank screen when loading in my browser. Developer console is empty under "Console" even with "Verbose" on.

In miniserv.error I see the following lines each time I loaded Webmin in my browser.

**Miniserv.error**

`24/Jun/2020:12:30:42 +0800] [nasa] / : Perl execution failed : Wide character in syswrite at /usr/libexec/webmin/miniserv.pl line 3140.`

__Ignore the host name 'nasa'. Its an enormous laptop I'm using and someone thought it would be amusing!_


When I view source for the blank page I see the following as the last few lines.

**Miniserv.log contents**

```
nasa - root [24/Jun/2020:12:25:05 +0800] "POST /session_login.cgi HTTP/1.1" 302 0
nasa - root [24/Jun/2020:12:25:06 +0800] "GET / HTTP/1.1" 500 120
nasa - root [24/Jun/2020:12:25:18 +0800] "GET / HTTP/1.1" 500 120
nasa - root [24/Jun/2020:12:25:18 +0800] "GET /unauthenticated/js/defaults.js?19519999999999901 HTTP/1.1" 200 5026
nasa - root [24/Jun/2020:12:25:18 +0800] "GET /images/favicons/webmin/favicon-32x32.png HTTP/1.1" 200 1787
nasa - root [24/Jun/2020:12:25:48 +0800] "GET /favicon.ico HTTP/1.1" 200 15086
nasa - root [24/Jun/2020:12:30:42 +0800] "GET / HTTP/1.1" 500 120
nasa - root [24/Jun/2020:12:30:42 +0800] "GET /unauthenticated/js/defaults.js?19519999999999901 HTTP/1.1" 200 5026
nasa - root [24/Jun/2020:12:30:42 +0800] "GET /images/favicons/webmin/favicon-32x32.png HTTP/1.1" 200 1787

```
**Webmin.log**

`1592972705.2348.0 [24/Jun/2020 12:25:05] craig ae73d9ca428c8084ed7d26099e2927b0 nasa global miniserv.pl "login" "-" "-"`
```
<script>HTTP/1.0 500 Perl execution failed
--
  | Server: MiniServ/1.94406231132
  | Date: Wed, 24 Jun 2020 04:25:17 GMT
  | Content-type: text/html; Charset=utf-8
  | Connection: close
  |  
  | <h1>Error - Perl execution failed</h1>
  | <p>Wide character in syswrite at /usr/libexec/webmin/miniserv.pl line 3140.
  | </p>

```
I hope this means something to someone, it's all dribble to me. :)

**Comment by swelljoe** [comment]
Yes, I guess it's related to the new translations @iliaross has merged, as it indicates a UTF-8 string without proper encoding. We may need @jcameron to help resolve it, though. Then again, it may already be fixed. They're both working on it heavily right now (which is why it can be dangerous to try to run the git version) and there may be a commit that fixes it between when you tried  to upgrade and it failed. Might be worth trying to upgrade again or look at the list of recent commits to see if there are any encoding related commits.

**Comment by Admiral-Noisy-Bottom** [comment]
Thanks for your help. I'll got take a look at recent commits.

Worst case scenario is that I rebuild the box. I make daily backups but it doesn't include /usr/libexec. 

Thanks again, I'll see what I can do.

**Comment by swelljoe** [comment]
That would be overkill.

You can reinstall the package using your package manager and it'll likely overwrite the offending files.

**Comment by Admiral-Noisy-Bottom** [comment]
Package Manager? I hope you don't mean the Package Manager in Webmin?  All I have access to is a command line via ssh or or at the server console.

Are you referring to yum? yum install webmin? 

I originally installed webmin via a script I got from the Webmin site.

**Comment by Admiral-Noisy-Bottom** [comment]
Just in case it is a theme issue, where are the themes located. I can see a folder called 'authentic-theme' in both /usr/libexec/webmin and /etc/webmin but they only contain 3 files.

If I knew where the authentic-theme files were stored I could restore them from this mornings backup.

**Comment by swelljoe** [comment]
3 files? Well, there's your problem right there. ;-)

The `/usr/libexec/webmin/authentic-theme/` folder should be chock full of goodness.

**Comment by Admiral-Noisy-Bottom** [comment]
Oh,sorry, I got it wrong. /usr/libexec/webmin/authentic-theme is full of stuff. I must have previously been in the wrong folder. Sorry to mislead you.

Making things worse :( I don't include the /usr folder in the daily backup. I include /etc /var/ /home and figured if I had to rebuild I could reinstall everything and copy back my config files.

Any idea how I change the theme to another theme, which hasn't been updated. There must be a user file somewhere which determines which theme each user has chosen.

**Comment by iliarostovtsev** [comment]
That is not something I observed when installing from repo.

What language do you have set?

What is the output of:

```
cat /etc/webmin/config
cat /etc/webmin/miniserv.conf
cat /etc/webmin/authentic-theme/favorites*
```

> That would be overkill. You can reinstall the package using your package manager and it'll likely overwrite the offending files.

Yes, you can easily revert to previous state by running `yum reinstall webmin`.

**Comment by Admiral-Noisy-Bottom** [comment]
@iliaross  Do you really want all the output of /etc/webmin/config and the other files you mentioned? That's a lot of data and is likely to run the internet out of space.

**Comment by iliarostovtsev** [comment]
> That's a lot of data and is likely to run the internet out of space.

:smile: Interesting. Usually that file is less than 50 lines. But yes, I want it! :wink:

**Comment by Admiral-Noisy-Bottom** [comment]
Files as requested below. Note there is no file or folder called favorites in /etc/webmin/authentic-theme folder

**miniserv.conf**

````
port=720
addtype_cgi=internal/cgi
realm=Webmin Server
logfile=/var/webmin/miniserv.log
errorlog=/var/webmin/miniserv.error
pidfile=/var/webmin/miniserv.pid
logtime=60
ssl=0
no_ssl2=1
no_ssl3=1
no_tls1=1
no_tls1_1=1
ssl_honorcipherorder=1
no_sslcompression=1
env_WEBMIN_CONFIG=/etc/webmin
env_WEBMIN_VAR=/var/webmin
atboot=1
logout=/etc/webmin/logout-flag
denyfile=\.pl$
log=1
blockhost_failures=3
blockhost_time=120
syslog=1
ipv6=0
session=1
userfile=/etc/webmin/miniserv.users
keyfile=/etc/webmin/letsencrypt-key.pem
passwd_file=/etc/shadow
passwd_uindex=0
passwd_pindex=1
passwd_cindex=2
passwd_mindex=4
passwd_mode=0
preroot=authentic-theme
passdelay=1
cipher_list_def=1
error_handler_403=403.cgi
error_handler_404=404.cgi
error_handler_401=401.cgi
nolog=\/stats\.cgi\?xhr\-stats\=general
sessiononly=/virtual-server/remote.cgi
preload=
logouttimes=root=10
no_resolv_myname=0
sockets=
extracas=/etc/webmin/letsencrypt-ca.pem
certfile=/etc/webmin/letsencrypt-cert.pem
perllib=
noshowstderr=
expires_paths=
gzip=
nolistdir=1
ssl_redirect=1
no_tls1_2=
preroot_root=authentic-theme
precache=none
logclear=1
logclf=0
loghost=1
expires=120
pam_conv=
blockuser_time=120
blockuser_failures=3
no_pam=1
logouttime=15
pam_end=
blocklock=
session_ip=
utmp=
failed_script=/etc/webmin/failed.pl
login_script=/etc/webmin/login.pl
logout_script=/etc/webmin/logout.pl
bind=
root=/usr/libexec/webmin
mimetypes=/usr/libexec/webmin/mime.types
server=MiniServ/1.94406231132
````

**config**

````
gotoone=1
acceptlang=0
os_version=15.8
passwd_file=/etc/shadow
tempdelete_days=7
theme_root=authentic-theme
help_width=
lang=en_GB.UTF-8
logsyslog=0
gotomodule=
logusers=
logemail=
logmodulesemail=
showhost=0
hostnamedisplay=
proxy_fallback=0
noremember=1
deftab=webmin
logfiles=1
logfullfiles=0
logtime=60
ld_env=LD_LIBRARY_PATH
path=/bin:/usr/bin:/sbin:/usr/sbin:/usr/local/bin
logclear=1
error_stack=0
proxy_pass=
passwd_cindex=2
realname_root=Craig
sclass=
real_os_type=CentOS Linux
mobile_theme=authentic-theme
sprio=
chattr=0
realname=
extra_headers=
syspath=
os_type=redhat-linux
tempdirdelete=1
hostnamemode=2
ld_path=
dateformat=dd/mon/yyyy
noproxy=
passwd_pindex=1
passwd_mindex=4
nofeedbackcc=1
referers_none=1
nomoduleup=
relative_redir=0
feedback_to=
passwd_uindex=0
real_os_version=7.8.2003
webprefix=
sysinfo=0
nowebminup=
by_view=0
log=1
logperms=
md5pass=1
theme=authentic-theme
nohostname=1
notabs=0
showlogin=0
cache_days=10
proxy_user=
sortconfigs=0
find_pid_command=ps auwwwx | grep NAME | grep -v grep | awk '{ print $2 }'
logmodules=
logsched=0
product=webmin
help_height=
````

**Comment by Admiral-Noisy-Bottom** [comment]
UPDATE

I have changed all mentions of 'authentic-theme' in /etc/config and attempted to login and now I have access to the old original webmin theme.

That tells me it is a theme issue rather than Webmin. :)

I'll continue trying to figure this out without the reinstall.

**Comment by iliarostovtsev** [comment]
Can you share content of `/etc/webmin/authentic-theme` and `/var/webmin/modules/authentic-theme`?

**Comment by Admiral-Noisy-Bottom** [comment]
/etc/webmin/authentic-theme

```
-rw-r--r--. 1 root root 4.3K Jun 24 15:11 settings.js
-rw-r--r--. 1 root root 2.2K Jun 24 15:11 settings-root
-rw-r--r--. 1 root root 127K Jun 24 15:11 stats-root.json
```
/var/webmin/modules/authentic-theme

```
-rw-r--r--. 1 root root 392 Jun 24 07:18 version-theme-development
-rw-r--r--. 1 root root 392 Jun 19 07:59 version-theme-stable

```

**Comment by iliarostovtsev** [comment]
Could you zip it up and attach it to this ticket (just drag&drop it to the editor).

**Comment by Admiral-Noisy-Bottom** [comment]
Sure, give me a second. I didn't know I could do that. :)

**Comment by Admiral-Noisy-Bottom** [comment]
[etc-webmin-authentic-theme.zip](https://github.com/webmin/webmin/files/4824743/etc-webmin-authentic-theme.zip)
[var-webmin-modules-authentic-theme.zip](https://github.com/webmin/webmin/files/4824745/var-webmin-modules-authentic-theme.zip)

**Comment by Admiral-Noisy-Bottom** [comment]
@iliaross 

FYI, I am currently in Webmin using the grey theme. I deleted authentic-theme via webmin configuration and reinstalled version 19.47 which I got from your git (https://github.com/authentic-theme/authentic-theme/releases/download/19.47/authentic-theme-19.47.wbt.gz)

I assigned authentic-theme to another user and tried to login to webmin. I got the same blank page.

**Comment by iliarostovtsev** [comment]
I unfortunately cannot reproduce it.

Is there a way to login and take a look?

**Comment by Admiral-Noisy-Bottom** [comment]
Did you want to log into webmin? Or ssh? I'm a bit funny about ssh and it is currently blocked. I can give you webmin though, temporarily.

**Comment by iliaross** [comment]
I would need SSH, otherwise it would be much slower.

You can add my pub key to **/root/.ssh/authorized_keys** file and make sure that **sshd_config** has `PermitRootLogin prohibit-password` directive set. You can privately email me details using my [iliaross.io](https://iliaross.io) website's contact form.

~~~
ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQClNgCrdnsYVpMu6Ne6OFecTsGighwB/TgdM7b1W3Z+BHFyXlWAsqjAzdcNyeMP3o18Xst8X3QT/13ly/F9+U+KFvD73mE3QuQUWjO6jmo9CDrf+LZxF0CR+z/HCgBJbxeuCt0/F/xkKTyJdIJaTMYSMIuPHNJXtsVYgOJ1Fvzsq9/0BDUw9lovgoQlhS+ywpheBal7UsB802DXKpmFJQ/l5WJtlxmhNy9zy6Oi483ADoRdwN+1U9YLaw5Rl3bTscLhPj8HBFrEUyk/oJn2TpKoqNOJ2Rw1sFL8mkxHokNam5GSybLosEW9g8sFgteOkht8+uRY9HZqWVbtT6hZCxJP Ilia

~~~

**Comment by Admiral-Noisy-Bottom** [comment]
Ok, give me a few minutes to set up port forwarding etc. I'll email you soon.

Thanks for your help, you have gone above and beyond the call of duty.

**Comment by iliarostovtsev** [comment]
I would like to ask you again - this is the issue you started seeing only after using installing Git version and if you reinstall using `yum reinstall webmin` you don't see this issue with any theme, including Authentic Theme, correct? Have you tried reinstalling? Have you tried running install script over again after reinstall?

**Comment by iliarostovtsev** [comment]
> Thanks for your help, you have gone above and beyond the call of duty.

I want to make sure that we don't get anything unexpected into production.

**Comment by Admiral-Noisy-Bottom** [comment]
I have not done yum resinstall yet, I thought that was a last resort.

Never had issues previously until I ran update-from-repo.sh. During the install it asked me if I wanted the new authentic-theme to which I said yes.

This all started because I clicked force update for the theme. I don't usually get advised of an updated theme. I got a big red warning that I needed a later version of webmin and to run update-from-repo.sh.

I ignored the warning about breaking my installation because I'm an idiot and don't take my own advice :)

Would you prefer I ran yum reinstall webmin first?

I didn't originally install via yum. I downloaded a script from the webmin site.

**Comment by iliarostovtsev** [comment]
> I thought that was a last resort.

Absolutely harmless and ordinary process.

> This all started because I clicked force update for the theme. I don't usually get advised of an updated theme. I got a big red warning that I needed a later version of webmin and to run update-from-repo.sh.

It's good that you noticed. In fact, if you're an English language user you could've ignored the warning and just update the theme without Webmin. I assume (as far as I remember) it would've just worked fine.

> I ignored the warning about breaking my installation because I'm an idiot and don't take my own advice :)

You can easily fix it by running reinstall command mentioned above.

> Would you prefer I ran yum reinstall Webmin first?

Well, if you can store the current system's state in a snapshot it would help, in case you run **./update-from-repo.sh** script again and it would work.

You know, as writing this, I think I know why this issue has happened. Do the following:

  1. Reinstall to previous state and make sure all is working just fine - `yum reinstall webmin`
  2. Then update upgrade update script with latest first - `curl https://raw.githubusercontent.com/webmin/webmin/master/update-from-repo.sh -o /usr/libexec/webmin/update-from-repo.sh`
  3. Then run `./update-from-repo.sh` script. 
  4. I think it will just work now, right?

**Comment by Admiral-Noisy-Bottom** [comment]
Ok, I'll do that and get back to you soon.

It's getting late at night here in Australia so I'll try and keep my eyes open for as long as I can.

**Comment by iliarostovtsev** [comment]
> It's getting late at night here in Australia so I'll try and keep my eyes open for as long as I can.

Thank you. I think you would be able to sleep nice after those 3 steps above.

**Comment by Admiral-Noisy-Bottom** [comment]
:) I'm not an expert in linux. By snapshot do you mean "Create Snapshot" within the Logical Volume Management?

**Comment by Admiral-Noisy-Bottom** [comment]
Ok, I did the yum reinstall and things are back to normal with authentic-theme working as it should.
I will now download the script, run it and get back to you soon.

**Comment by Admiral-Noisy-Bottom** [comment]
Bad news!

After running yum reinstall webmin everything was back to normal and working as expected.

I downloaded the update script and ran it to completion.

Back to where I started. I get the login prompt and then a blank page. I'll reinstall and leave it as it was. Am happy to try anything else for you to iron out bugs before you release the update.  

I'm afraid I'm going to have to give up for tonight, I'm sorry.

But you have been awesome and your support is much appreciated.

**Comment by iliarostovtsev** [comment]
> I'll reinstall and leave it as it was. Am happy to try anything else for you to iron out bugs before you release the update.

Okay, try the last super simple way - 

~~~
yum update https://download.webmin.com/devel/rpm/webmin-1.943-1.noarch.rpm
~~~

If that doesn't work I'm gonna get worried.

If it works, afterwards, update the theme to latest dev version from theme configuration page.

**Comment by Admiral-Noisy-Bottom** [comment]
Ok, I'm doing it now.

**Comment by iliarostovtsev** [comment]
Thank you!

**Comment by Admiral-Noisy-Bottom** [comment]
Still have a blank screen.

I cannot now run yum reinstall webmin

I get the following error;

`Installed package webmin-1.943-1.noarch (from /webmin-1.943-1.noarch) not available`

How do I go back?

**Comment by Admiral-Noisy-Bottom** [comment]
See my previous message about not being able to reinstall.

I've got to get some sleep.  Again, thank you so much for all your help.

**Comment by iliarostovtsev** [comment]
> See my previous message about not being able to reinstall.

Simply run `yum downgrade webmin` command.

Okay, then please send me credentials for logging in and having a look. It's odd!

**Comment by iliarostovtsev** [comment]
If you have it un-cleared yet, screenshot and paste here the output of running `./update-from-repo.sh` script.

**Comment by Admiral-Noisy-Bottom** [comment]
@iliaross I've emailed you via your site.

**Comment by iliarostovtsev** [comment]
Okay, I logged it and fixed an issue for you with latest upgrade and enabling SSL. I could reproduce an issue on my system with disabled SSL. I will have to take a closer look why it's happeing.

**Comment by iliarostovtsev** [comment]
I am gonna need your help on it. I would like to point out this issue only happens when data sent to `syswrite` but not to `Net::SSLeay::write` (i.e. `ssl=0` in **miniserv.conf**). If using `binmode(SOCK, ":utf8");` and `binmode(FILE, ":utf8");` then it seems to solve the problem but then it double encodes data and all Unicode chars look broken.

@jcameron @swelljoe any idea how to fix it?

**Comment by jcameron** [comment]
So is this issue only triggered in non-SSL mode, when using en_GB.UTF-8 as the language?

**Comment by jcameron** [comment]
Wait a second, en_GB isn't even one of our supported languages.  What if you edit `/etc/webmin/config` and remove the `lang=en_GB.UTF-8` line?

**Comment by iliarostovtsev** [comment]
No, no. Language doesn't matter. It's all about `syswrite` choking on chars greater than 255.

**Comment by jcameron** [comment]
Ok, but that code hasn't changed for ages - why is it suddenly an issue now? I suspect this is related to some recent language or charset changes.

**Comment by iliarostovtsev** [comment]
By the way, all old type languages are processed by `sub safe_language {}`.

In-fact en-GB was missing on safe language map. [Fixed here](https://github.com/webmin/webmin/commit/e5ee0df57405efa01203f995b295e5c90f176480).

**Comment by iliarostovtsev** [comment]
>  I suspect this is related to some recent language or charset changes.

Yes, it is.

https://github.com/webmin/webmin/commit/adb4ad1797cdd37a87eacaa0755edaad93a8c182#diff-8b0d1e2d31f69c1826d4ecc01adb23f1

**Comment by iliarostovtsev** [comment]
However, that particular change (**iso-8859-1 -> utf-8**) doesn't change anything in this regard.

**Comment by iliarostovtsev** [comment]
Accidentally closed it. Re-opening..

**Comment by jcameron** [comment]
Possible fix : https://github.com/webmin/webmin/commit/c783eca0a2d5b0f41092f8945559601a5e066a03

**Comment by iliarostovtsev** [comment]
Finishing work on this.

https://github.com/webmin/webmin/commit/e712efb31f14494de8c6133c8893e17ff8fd9ef8
